### PR TITLE
Add episode 18

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/18.md [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+Now, when you want to run any further analysis on this intermediate dataset, you
+simply refer to the view instead of having to rewrite the statement before. As
+mentioned, this view also has a schema and is treated much like a table when the
+query engine does its planning. In this way it is also easier to map the data to
+the application logic by enabling different shapes of the same data.
+
+Another reason why you would want to create a view is to control access to the
+data. When running the query, you get to choose which columns and rows get
+filtered out and that return from when users query the view. The authorization 
+of a user is only to read the abstraction of the data rather than the real data.
```
Comment (mosabua) [inline _episodes/18.md:158]: of a use is tied to the view and its content, and that can significantly differ from the complete data in the underlying tables. For example, the views can exclude sensitive data like social security numbers, birth dates, credit card numbers, and many other facts.

File: _episodes/18.md [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+
+When creating a view, there are two modes that the view can run in that will 
+indicate the user that will run the queries defined in the view during query
+runtime. You can either run this query as the `DEFINER` which indicates to run
+the view query as the user that created the view, or as the `INVOKER`, which
+indicates to run the view query as the user that is running the outer query of
+the view. The default mode is `DEFINER`. See more 
+[here](https://trino.io/docs/current/sql/create-view.html#security).
```
Comment (mosabua) [inline _episodes/18.md:166]: replace "here" with "in the security section of the create view documentation".

File: _episodes/18.md [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+
+There are two types of views; materialized and logical views. The view defined
+above is the standard logical view that gets expanded into its definition. 
+Logical views don't provide any performance benefit since the data is not stored
+upfront and instead queried at query time. Materialized views persist the view
```
Comment (mosabua) [inline _episodes/18.md:171]: remove "upfront"

File: _episodes/18.md [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+the view query as the user that created the view, or as the `INVOKER`, which
+indicates to run the view query as the user that is running the outer query of
+the view. The default mode is `DEFINER`. See more 
+[here](https://trino.io/docs/current/sql/create-view.html#security).
+
+There are two types of views; materialized and logical views. The view defined
+above is the standard logical view that gets expanded into its definition. 
+Logical views don't provide any performance benefit since the data is not stored
+upfront and instead queried at query time. Materialized views persist the view 
+data upon view creation by storing the query data on disk.
```
Comment (mosabua) [inline _episodes/18.md:172]: upon creation by storing the query data. 

might be on disk or not...

File: _episodes/18.md [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+
+Schemas for logical views can't be quickly referenced as was done for the table 
+using `SHOW CREATE TABLE`. The query engine can quickly derive each of these
+types. The `SHOW CREATE VIEW` only shows the original query that was used to
```
Comment (mosabua) [inline _episodes/18.md:176]: this sentence is confusing..

File: _episodes/18.md [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+data upon view creation by storing the query data on disk.
+
+Schemas for logical views can't be quickly referenced as was done for the table 
+using `SHOW CREATE TABLE`. The query engine can quickly derive each of these
+types. The `SHOW CREATE VIEW` only shows the original query that was used to 
+create the view for logical tables but will show the actual schema of the
+materialized data.
+
+Materialized views make overall queries much faster to run as part of the query
+has already been computed and all that is required is recall. One issue with
```
Comment (mosabua) [inline _episodes/18.md:181]: remove "and all that is required is recall"

File: _episodes/18.md [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+
+Schemas for logical views can't be quickly referenced as was done for the table 
+using `SHOW CREATE TABLE`. The query engine can quickly derive each of these
+types. The `SHOW CREATE VIEW` only shows the original query that was used to 
+create the view for logical tables but will show the actual schema of the
+materialized data.
+
+Materialized views make overall queries much faster to run as part of the query
+has already been computed and all that is required is recall. One issue with 
+materialized views is that the data may become outdated or new data may become
```
Comment (mosabua) [inline _episodes/18.md:182]: that the data can become outdated and out of sync with the underlying table data. To keep the data between the tables and materialized view in sync, you have to refresh the view.

File: _episodes/18.md [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+available in the original tables. These tables can periodically be updated using
+a special refresh command or periodically have this procedure run automatically. 
+
+### Trino Views: So many views, so little time
```
Comment (mosabua) [inline _episodes/18.md:186]: Trino views

File: _episodes/18.md [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+
+Views handling in Trino depends on the connector. In general, most connectors
+expose views to Trino as if they are another set of tables available for Trino
+to query. The main exceptions for this is the Hive and Iceberg connectors, which
```
Comment (mosabua) [inline _episodes/18.md:190]: connectors. 

Remove the rest of this sentence.

File: _episodes/18.md [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+require Trino to do more as Trino is the primary query engine for these
+connectors. This simplifies thinking about how Trino handles views in general.
```
Comment (mosabua) [inline _episodes/18.md:192]: Remove the sentence "This simplifies... "

File: _episodes/18.md [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+  </tr>
+  <tr>
+    <td colspan="2">Hive Created View</td>
+    <td>✅ (read-only)</td>
+    <td>✅ (read-only)</td>
+  </tr>
+</tbody>
+</table>
+
+You'll notice that the Materialized views cannot be created through the Hive
```
Comment (mosabua) [inline _episodes/18.md:223]: materialized

File: _episodes/18.md [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+```
+Also, you cannot create logical views in Iceberg and you will get the following
+exception:
+
+```
+Caused by: java.sql.SQLException: Query failed (#...): 
+This connector does not support creating views.
+```
+
+#### Trino Reads Hive Views
```
Comment (mosabua) [inline _episodes/18.md:238]: Trino read Hive views

File: _episodes/18.md [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+
+Before Trino there was Hive. For many, Trino was the replacement for their Hive
```
Comment (mosabua) [inline _episodes/18.md:240]: Trino is a replacement for the Hive runtime for many users, and it is very useful for these users to also be able to read data from Hive views in Trino.

File: _episodes/18.md [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+you will gradually migrate these views to Trino over time. Trino also supports
+reading Hive materialized views, though Trino reads these views as another Hive 
+table rather since they are stored similarly to standard Hive tables. Since
+Hive views are defined in HiveQL, the view definitions need to be translated to
+Trino SQL syntax. This is done using LinkedIn's Coral library.
+
+#### Coral: the unifier of the bee and the bunny
+
+[Coral](https://github.com/linkedin/coral) is a project that allows for 
+translation between views from different SQL syntax. It takes in SQL from Hive
```
Comment (mosabua) [inline _episodes/18.md:255]: It can process Hive QL statements and convert them to an internal

File: _episodes/18.md [hunk 15]
```diff
@@ -0,0 +1,560 @@
...
+that are in sync with latest data.
+
+Automatically keeping materialized views fresh can be tricky from resource 
+management point of view since the computation to materialize the materialized 
+view can be expensive. Trino currently does not support automatic refresh of 
+materialized views. It instead supports the `REFRESH MATERIALIZED VIEW` command 
+that the user can issue to ensure that the materialized view is fresh.
+
+As a part of executing `REFRESH MATERIALIZED VIEW` command in Trino, existing
+data in the materialized view drops and new/fresh data re-inserted if there have been any
```
Comment (mosabua) [inline _episodes/18.md:314]: is dropped and new data is inserted

File: _episodes/18.md [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+
+### Support in Trino and at Netflix:
+
+#### Netflix materialized views
+
+The main reason Netflix is interested in materialized views is to give analysts 
+an easy way to compute and materialize their frequently used queries and keep 
+the results refreshed without relying on ETL pipeline to create and maintain 
+those result sets. Some materialized views are as simple as queries that project
+columns and apply filters, selecting data for a time range or for a test-id,
```
Comment (mosabua) [inline _episodes/18.md:344]: test-id. Others are

File: _episodes/18.md [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+an easy way to compute and materialize their frequently used queries and keep 
+the results refreshed without relying on ETL pipeline to create and maintain 
+those result sets. Some materialized views are as simple as queries that project
+columns and apply filters, selecting data for a time range or for a test-id, 
+others are more complex that perform multi-level joins and aggregations.
+
+#### Netflix materialized view cross compatibility extension 
+
+Materialized views, much like logical views are compatible across the two main 
+engines used at Netflix, Trino and Spark. A little more about how Netflix uses
```
Comment (mosabua) [inline _episodes/18.md:350]: A little more...

thats not a sentence.. reword please

File: _episodes/18.md [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+those result sets. Some materialized views are as simple as queries that project
+columns and apply filters, selecting data for a time range or for a test-id, 
+others are more complex that perform multi-level joins and aggregations.
+
+#### Netflix materialized view cross compatibility extension 
+
+Materialized views, much like logical views are compatible across the two main 
+engines used at Netflix, Trino and Spark. A little more about how Netflix uses 
+Trino to understand why logical views and materialized views are important. 
+Spark is used at Netflix to do ETL, creating and populating tables, while Trino
```
Comment (mosabua) [inline _episodes/18.md:352]: break up into multiple sentences

File: _episodes/18.md [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+#### Netflix materialized view cross compatibility extension 
+
+Materialized views, much like logical views are compatible across the two main 
+engines used at Netflix, Trino and Spark. A little more about how Netflix uses 
+Trino to understand why logical views and materialized views are important. 
+Spark is used at Netflix to do ETL, creating and populating tables, while Trino 
+is the most popular engine with analysts and developers for adhoc/experimental 
+queries as well as audits. 
+
+Trino is also used for CTAS in some use cases. Both the engines access data from
```
Comment (mosabua) [inline _episodes/18.md:356]: is also use to insert data with CREATE TABLE AS SELECT (CTAS) statements in some use cases

File: _episodes/18.md [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+tables using Iceberg and Hive connectors where data is stored in S3. Netflix 
+built upon the Trino logical views to create common views that are accessible 
+from both Spark and Trino. The difference between the Trino logical views and 
+Netflix common views is that the metadata is stored in the hive metastore for
```
Comment (mosabua) [inline _episodes/18.md:360]: Hive

File: _episodes/18.md [hunk 21]
```diff
@@ -0,0 +1,560 @@
...
+Spark is used at Netflix to do ETL, creating and populating tables, while Trino 
+is the most popular engine with analysts and developers for adhoc/experimental 
+queries as well as audits. 
+
+Trino is also used for CTAS in some use cases. Both the engines access data from
+tables using Iceberg and Hive connectors where data is stored in S3. Netflix 
+built upon the Trino logical views to create common views that are accessible 
+from both Spark and Trino. The difference between the Trino logical views and 
+Netflix common views is that the metadata is stored in the hive metastore for 
+Trino logical views, while common views store their metadata in json format in S3.
```
Comment (mosabua) [inline _episodes/18.md:361]: JSON

File: _episodes/18.md [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+queries as well as audits. 
+
+Trino is also used for CTAS in some use cases. Both the engines access data from
+tables using Iceberg and Hive connectors where data is stored in S3. Netflix 
+built upon the Trino logical views to create common views that are accessible 
+from both Spark and Trino. The difference between the Trino logical views and 
+Netflix common views is that the metadata is stored in the hive metastore for 
+Trino logical views, while common views store their metadata in json format in S3. 
+
+A view object in hive metastore points to the S3 location of metadata. It tracks
```
Comment (mosabua) [inline _episodes/18.md:363]: Hive

File: _episodes/18.md [hunk 23]
```diff
@@ -0,0 +1,560 @@
...
+evolution of view definition in the form of versions so that one can potentially
```
Comment (mosabua) [inline _episodes/18.md:364]: so that you can

File: _episodes/18.md [hunk 24]
```diff
@@ -0,0 +1,560 @@
...
+revert a view to its older version. Main benefit of common views is 
+interoperability between Spark and Trino (can create, replace, query, drop from 
+either engine and can be expanded to other engines). Netflix supports common 
+views through both Hive and Iceberg connectors. 
+
+Currently, common views support SQL syntax common to both Spark and Trino. This 
+support can be expanded in future using linkedIn’s coral project such that
```
Comment (mosabua) [inline _episodes/18.md:371]: LinkedIn's Coral

**Review by mosabua — CHANGES_REQUESTED** [review]
A bunch of changes necessary. After that probably good to go.

File: _episodes/18.md (hunk 5 above)
Comment (bitsondatadev) [reply _episodes/18.md:176]: Removing...it doesn't add much and I found out the point I was trying to make isn't true in all cases :)

**Review by mosabua — APPROVED** [review]
Lets ship this as it stands to get it live

**Pull request merged**